[Deprecated fork PR] Rectangular convolutions support

curiousdragon opened the pull request:
- Rectangular convolution tests can be found by running `./build.sh` in `gemmini-rocc-tests`, and will end with `-rect-baremetal`
- Current failing tests:
  - `conv_dw-rect-baremetal`
  - `conv_trans_input_3120_with_kernel_dilation-rect-baremetal`
  - `conv_with_pool-rect-baremetal`
- I suspect this might be related with `gemmini_extended2_config_st`, called in `sp_tiled_conv`, since now the `pool_out_dim` has been split into `pool_out_row_dim` and `pool_out_col_dim`, but `gemmini_extended2_config_st` seems to be a ROCC instruction so it's unclear to me how exactly to proceed. (Or the bug might be somewhere else, I have yet to investigate further)

curiousdragon committed:
Setup for rectangular convolutions

curiousdragon committed:
Update main convolution logic for rectangular option

curiousdragon committed:
Add example test for rectangular conv, update Makefile

curiousdragon committed:
Add RECT flag for building tests

hngenc commented on Makefile.in:
```diff
@@ -43,6 +43,16 @@ test-baremetal-bareMetalC:
 		RISCVTOOLS=$(RISCVTOOLS) \
 		RUNNER=$(RUNNER) \
 		run-baremetal
+	make -C bareMetalC	\
+	        -f $(abs_top_srcdir)/bareMetalC/Makefile \
+                TARGET_MAKEFILE=$(abs_top_srcdir)/bareMetalC/Makefile \
+		abs_top_srcdir=$(abs_top_srcdir) \
+	 	src_dir=$(abs_top_srcdir)/bareMetalC \
+	 	XLEN=$(XLEN) \
+	 	PREFIX=$(ROCC)-bareMetalC \
+		RISCVTOOLS=$(RISCVTOOLS) \
+		RUNNER=$(RUNNER) \
+		run-rect-baremetal
```
I'd rather that we leave this out of the PR. We just just include the `run-rect-baremetal` tests with the rest of the baremetal tests.

hngenc commented on bareMetalC/conv.c:
```diff
@@ -22,13 +23,25 @@
 #else
 
-#define IN_DIM 17
+#ifdef RECT
```
Let's get rid of the `RECT` macro and instead create a new `conv_rect.c` file

hngenc commented on include/gemmini.h:
```diff
@@ -349,11 +349,25 @@ static void counter_reset() {
     ROCC_INSTRUCTION_RS1_RS2(XCUSTOM_ACC, ((uint64_t)(weightA) << 8) | ((low_D) << 2) | ((full_C) << 1) | (ex_accumulate), ((B_transpose) << 1) | (A_transpose), k_LOOP_WS) \
   }
 
+// jamie: look at this later
```
Can we remove all the comments with "jamie" in them?

hngenc commented on include/gemmini.h:
```diff
@@ -349,11 +349,25 @@ static void counter_reset() {
+// k - kernel rows, cols, channels
+// l, r, u, d pad - left, right, up, down padding
+// pl, pr, pu, pd pool - pool left, pool right, pool up, pool down (?)
+// o - output rows, cols
```
Let's get rid of these commented lines. We document the exact syntax in the Scala files anyways, and I'd rather not do that in two separate places.

hngenc commented on include/gemmini.h:
```diff
@@ -1215,9 +1230,10 @@ static void sp_tiled_conv(
   }
 
-  gemmini_loop_conv_ws(batch_size, in_dim, in_channels, out_channels, out_dim, pool_out_dim, stride, padding, kernel_dim, kernel_dilation, pool_size, pool_stride, pool_padding, batches, porows, pocols, pochs, krows, kcols, kchs, lpad, rpad, upad, dpad, plpad, prpad, pupad, pdpad, orows, ocols, weights, output, bias, input, no_bias, no_pool, downsample, wrot180, input_dilated, act, trans_output_12 ...
+  // gemmini_loop_conv_ws(batch_size, in_row_dim, in_col_dim, in_channels, out_channels, out_row_dim, out_col_dim, pool_out_row_dim, pool_out_col_dim, stride, padding, kernel_dim, kernel_dilation, pool_size, pool_stride, pool_padding, batches, porows, pocols, pochs, krows, kcols, kchs, lpad, rpad, upad, dpad, plpad, prpad, pupad, pdpad, orows, ocols, weights, output, bias, input, no_bias, no_pool ...
```
Instead of commenting out the `gemmini_loop_conv_ws` call, let's add an if-statement which calls `gemmini_loop_conv_ws` if we're doing a square conv, but uses your new C code if we're doing a rectangular conv.

hngenc reviewed (changes requested):
This looks great overall! I have a few requests though:

1. Can you try creating a branch on Gemmini as well, and bumping the `gemmini-rocc-tests` submodule on it so that CI runs?
2. For the convs which don't work in the rectangular case (like `conv_with_pool-rect`), can we add if-statements in `tiled_conv_auto` which error out if we don't think will work (e.g. if they use max-pooling).

curiousdragon committed:
Fix trans_input_3120 conv dim indexing logic

curiousdragon committed:
Add check for rect convs, no support for rect pool

Additional changes:
- Remove extra comments
- Add explanation and FAILED exit for rect convs with pooling

curiousdragon committed:
Fix dimensions in conv_dw_cpu_without_pool

curiousdragon committed:
Add square check in tiled_conv_downsample

This function seems to only be called for resnet50,
which uses square dimensional convolutions. Since there doesn't seem
to be a test for calling this function with rectangular (i.e.
non-square) dimensions, I've currently added a check that the
dimensions this function is being called with are square dimensions.

curiousdragon commented:
Update: I actually ended up finding and fixing a few dimensional bugs, which seemed to be the source of the errors in
- `conv_dw-rect-baremetal` and
- `conv_trans_input_3120_with_kernel_dilation-rect-baremetal`

so these two tests are now passing!

Now, the only rectangular convolution test that is failing is `conv_with_pool-rect-baremetal`, and as suggested, I've added an `if` statement to exit with an error message and fail if someone tries to run a rectangular convolution with pooling - they would get something along the lines of:
```bash
Pooling with rectangular convolutions is currently not supported.
*** FAILED *** (tohost = 1)
```
I've also added this error message and exit statement to the commented out pooling logic in `sp_tiled_conv` and `sp_tiled_conv_dw` just in case, since I don't think the `gemmini_extended2_config_st` function currently supports differing pooling output dimensions which rectangular convolutions probably require.

Remaining todos (edit: done in #28):
- [x] adjust the makefile
- [x] fix merge conflicts with the current main branch

hngenc commented:
Thanks for the revisions!

curiousdragon commented:
Note: Closing this PR since it's based on a branch in a fork, and opening up a new PR (#28) that has an updated branch on this repository specifically for review + merging! If I remember correctly, I think the main reasoning for wanting to have a non-forked development branch was that now with #28, I can make a PR for the `gemmini` repository with an updated submodule reference that isn't to a separate fork of this `gemmini-rocc-tests` repository.
